core/src/main/scala/kafka/log/Log.scala

```diff
@@ -549,19 +569,31 @@ class Log(@volatile var dir: File,
         // update the PID sequence mapping
         for ((pid, producerAppendInfo) <- appendInfo.producerAppendInfos) {
           trace(s"Updating pid with sequence: $pid -> ${producerAppendInfo.lastEntry}")
-
           if (assignOffsets)
             producerAppendInfo.assignLastOffsetAndTimestamp(appendInfo.lastOffset, appendInfo.maxTimestamp)
           pidMap.update(producerAppendInfo)
         }
 
+        // update the transaction index and last stable offset. The ordering is important.
+        // The index must be updated before we can advance the LSO, but the LSO value must take
+        // into account the completed transaction.
+        for (completedTxn <- appendInfo.completedTransactions) {
+          val firstUnstableOffset = pidMap.firstUndecidedOffsetExcluding(completedTxn)
+          val lastStableOffset = firstUnstableOffset.getOrElse(completedTxn.lastOffset + 1)
+          segment.updateTxnIndex(completedTxn, lastStableOffset)
+          pidMap.completeTxn(completedTxn)
+        }
+
         // always update the last pid map offset so that the snapshot reflects the current offset
         // even if there isn't any idempotent data being written
         pidMap.updateMapEndOffset(appendInfo.lastOffset + 1)
 
         // increment the log end offset
         updateLogEndOffset(appendInfo.lastOffset + 1)
 
+        // update the first unstable offset (which is used to compute LSO)
+        updateFirstUnstableOffset()
```

Comment:
Hmm... good point. I had thought it was necessary since the first unstable offset could be lower than the high watermark, but we wouldn't be able to advance it any further from this write because any appended transaction markers could not have been replicated yet.